### PR TITLE
Fix batch_get_for_update

### PR DESCRIPTION
**Pull request by ekexium**
The PR tries to fix the `batch_get_for_update` by introducing a new plan `PreserveKeys` and a merge strategy `MergeAndMatchKey`.

### CODE REVIEW

File: src/request/plan_builder.rs [hunk 1]
```diff
@@ -14,6 +15,8 @@ use crate::{
 use std::{marker::PhantomData, sync::Arc};
 use tikv_client_store::HasError;
 
+use super::plan::PreserveKey;
```
Comment (nrc) [inline src/request/plan_builder.rs:18]: add this to the crate import

File: src/request/shard.rs [hunk 2]
```diff
@@ -9,6 +10,8 @@ use crate::{
 use futures::stream::BoxStream;
 use std::sync::Arc;
 
+use super::plan::PreserveKey;
```
Comment (nrc) [inline src/request/shard.rs:13]: and here

File: src/request/shard.rs [hunk 3]
```diff
@@ -51,6 +54,21 @@ impl<P: Plan + Shardable, PdC: PdClient> Shardable for ResolveLock<P, PdC> {
...
+        pd_client: &Arc<impl PdClient>,
+    ) -> BoxStream<'static, Result<(Self::Shard, Store)>> {
+        self.inner.shards(pd_client)
+    }
+
+    fn apply_shard(&mut self, shard: Self::Shard, store: &Store) -> Result<()> {
+        self.inner.apply_shard(shard, store)
+    }
+}
+
```
Comment (nrc) [inline src/request/shard.rs:71]: Could you use an impl to generate this impl and the RetryRegion one?

File: src/kv/key.rs [hunk 4]
```diff
@@ -128,6 +128,10 @@ impl Key {
     }
 }
 
+pub trait HasKeys {
```
Comment (nrc) [inline src/kv/key.rs:131]: I think it is better for this trait to live in src/requests/plan.rs

File: src/request/plan.rs [hunk 5]
```diff
@@ -292,6 +313,38 @@ where
...
+impl<P> Plan for PreserveKey<P>
+where
+    P: Plan + HasKeys,
+{
+    type Result = ResponseAndKeys<P::Result>;
+
+    async fn execute(&self) -> Result<Self::Result> {
+        let keys = self.inner.get_keys();
+        let res = self.inner.execute().await?;
+        // TODO: should we check they have the same length?
```
Comment (nrc) [inline src/request/plan.rs:343]: I don't think we can, since `res` may not be a list

File: src/request/plan.rs [hunk 6]
```diff
@@ -349,3 +402,60 @@ mod test {
             .for_each(|r| assert!(r.is_err()));
     }
 }
+
```
Comment (nrc) [inline src/request/plan.rs:405]: These items should come before the tests in the file

File: src/request/plan.rs [hunk 7]
```diff
@@ -349,3 +402,60 @@ mod test {
...
+// contains a response and the corresponding keys
+// currently only used for matching keys and values in pessimistic lock requests
+#[derive(Debug, Clone)]
+pub struct ResponseAndKeys<Resp>(Resp, Vec<Key>);
```
Comment (nrc) [inline src/request/plan.rs:409]: Could this just be a tuple?

**Review by nrc — COMMENTED** [review]
Mostly looks good, a few comments but they're mostly about style.

File: src/request/plan.rs (hunk 7 above)
Comment (ekexium) [reply src/request/plan.rs:409]: I doubt tuples can be ambiguous. It's not intuitive to tell what `(Response, Vec<Key>)` means, IMO.

File: src/request/shard.rs (hunk 3 above)
Comment (nrc) [reply src/request/shard.rs:71]: whoops, _macro_ to generate the impl

File: src/request/plan.rs (hunk 7 above)
Comment (nrc) [reply src/request/plan.rs:409]: IMO the reading of the tuple is exactly `Response` and `Key`s. To be a bit abstract, I believe that a tuple struct is equivalent to a newtype and similarly should be used if and only if the newtype introduces an invariant over the component types.

Comment (ekexium) [reply src/request/plan.rs:409]: Makes sense. 
I think one thing that introducing the new type benefits is that the traits are implemented only for this type but not for general `(Resnponse, Vec<Key>)`. So it won't interfere with other uses of `(Resnponse, Vec<Key>)` if we use it for other purposes, or even where we may need different implementations of these traits (though not very likely).

Comment (nrc) [reply src/request/plan.rs:353]: Are there other uses of (Response, Vec<Key>) where this impl does not make sense?

Comment (ekexium) [reply src/request/plan.rs:353]: I can't think of one...

**Comment by nrc** [comment]
ping @ekexium what's the status of this PR?

**Comment by ekexium** [comment]
> ping @ekexium what's the status of this PR?

I've been busy in other things. Will update it soon.

**Comment by ekexium** [comment]
@nrc PTAL, I think it's almost done. Based on the discussion on Slack, we keep the `ResponseAndKeys`.

**Review by nrc — APPROVED** [review]
lgtm

**Pull request merged**